03-rbac/proposal.md

```diff
@@ -0,0 +1,159 @@
+# Summary
+
+Introduces a proposal for RBAC. Specifically:
+
+- What roles do we need?
+- How would fly set-team work?
+- How do roles get persisted in the database?
+- How are roles handled by the API?
+
+
+# Proposal
+
+## What roles do we need?
+
+`global_admin` or `admin` - Granted if you are part of an admin team. Currently only the `main` team is an admin team.
+
+`team_member` or `member` - Granted during `fly set-team`. Has read/write access to all the team's resources.
+
+`team_viewer` or `viewer` - Granted during `fly set-team`. Has readonly access to all the team's resources.
+
+`public` - Not a role. This applies to pipelines. Public pipelines will behave as they did before.
+
+
+## How would fly set-team work?
+
+Currently you fly set-team like the following:
+```bash
+fly -t mytarget set-team -n myteam --allow-all-users
+fly -t mytarget set-team -n myotherteam --github-user pivotal-jwinters --github-team myorg:myteam
+```
+
+We could allow roles to be set as follows:
+```bash
+fly -t mytarget set-team -n myteam --role viewer --allow-all-users
+fly -t mytarget set-team -n myteam --role member --github-user pivotal-jwinters --github-team myorg:myteam
+```
+
+The main drawback here is that we would have to merge `auth` configuration because `viewers` and `members` would be configured in two separate steps.  The current behaviour overwrites the entire auth config each time, so in keeping with the current behaviour I guess we would overwrite the entire config for each role separately. 
+
+For backwards compatibility, the default role would be `member` if you don't pass the `--role` flag.
+
+
+#### Maybe time for config files?
+
+An alternative approach would be to start introducing config files for `fly set-team(s)`.
+
+```bash
+fly -t mytarget set-teams -c /tmp/team-config
+```
+
+Where `/tmp/team-config` might look something like:
+
+```yaml
+teams: 
+  main: 
+    admin: true
```

Comment:
Should this be configurable for the `main` team? What's the recovery if someone makes this false?